### PR TITLE
nifm: add various structs

### PR DESCRIPTION
**Pull request by VolcaEM**
The structs are taken from Switchbrew (https://switchbrew.org/wiki/Network_Interface_services)
Error codes taken from Ryujinx (https://github.com/Ryujinx/Ryujinx), this file in particular: https://github.com/Ryujinx/Ryujinx/blob/master/Ryujinx.HLE/HOS/Services/Nifm/ResultCode.cs

### CODE REVIEW

File: src/core/hle/service/nifm/nifm.cpp [hunk 1]
```diff
@@ -217,7 +217,17 @@ class IGeneralService final : public ServiceFramework<IGeneralService> {
             rb.Push<u8>(1);
         }
     }
+    void GetCurrentIpAddress(Kernel::HLERequestContext& ctx) {
+        LOG_WARNING(Service_NIFM, "(STUBBED) called");
+        
+        auto const current_ip_address = network_profile_data.ip_data.address_settings.ip_address.address;
```
Comment (FearlessTobi) [inline src/core/hle/service/nifm/nifm.cpp:223]: I'm sure just passing an empty IP address can cause issues down the line and shouldn't be done.

File: src/core/hle/service/nifm/nifm.cpp [hunk 2]
```diff
@@ -13,6 +13,9 @@
 
 namespace Service::NIFM {
 
+constexpr ResultCode ERR_NO_INTERNET_CONNECTION = {ErrorModule::NIFM, 0x12C};
```
Comment (lioncash) [inline src/core/hle/service/nifm/nifm.cpp:16]:
```suggestion
constexpr ResultCode ERR_NO_INTERNET_CONNECTION{ErrorModule::NIFM, 0x12C};
```

File: src/core/hle/service/nifm/nifm.cpp [hunk 3]
```diff
@@ -13,6 +13,9 @@
...
+constexpr ResultCode ERR_OBJECT_IS_NULL = {ErrorModule::NIFM, 0x15E};
```
Comment (lioncash) [inline src/core/hle/service/nifm/nifm.cpp:17]:
```suggestion
constexpr ResultCode ERR_OBJECT_IS_NULL{ErrorModule::NIFM, 0x15E};
```

File: src/core/hle/service/nifm/nifm.cpp [hunk 4]
```diff
@@ -217,7 +220,22 @@ class IGeneralService final : public ServiceFramework<IGeneralService> {
             rb.Push<u8>(1);
         }
     }
+    void GetCurrentIpAddress(Kernel::HLERequestContext& ctx) {
+        LOG_WARNING(Service_NIFM, "(STUBBED) called");
+
+        auto const current_ip_address =
```
Comment (lioncash) [inline src/core/hle/service/nifm/nifm.cpp:226]:
```suggestion
        const auto current_ip_address =
```

File: src/core/hle/service/nifm/nifm.h [hunk 5]
```diff
@@ -14,6 +14,140 @@ class System;
 
 namespace Service::NIFM {
 
+#pragma pack(1)
```
Comment (lioncash) [inline src/core/hle/service/nifm/nifm.h:17]: This is unnecessary. The struct only contains a 4 byte member, so there will be no padding to get rid of.

File: src/core/hle/service/nifm/nifm.h [hunk 6]
```diff
@@ -14,6 +14,140 @@ class System;
...
+struct ClientId {
+    u32 client_id{};
+};
+#pragma pack()
+static_assert(sizeof(ClientId) == 0x4, "ClientId has incorrect size.");
+
+#pragma pack(1)
```
Comment (lioncash) [inline src/core/hle/service/nifm/nifm.h:24]: Ditto

File: src/core/hle/service/nifm/nifm.h [hunk 7]
```diff
@@ -14,6 +14,140 @@ class System;
...
+struct Uuid {
+    u128 uuid{};
+};
+#pragma pack()
+static_assert(sizeof(Uuid) == 0x10, "Uuid has incorrect size.");
+
+#pragma pack(1)
```
Comment (lioncash) [inline src/core/hle/service/nifm/nifm.h:31]: Ditto, this isn't necessary

File: src/core/hle/service/nifm/nifm.h [hunk 8]
```diff
@@ -14,6 +14,140 @@ class System;
...
+    std::array<u8, 0x20> ssid{};
+    u8 unk1{};
+    u8 unk2{};
+    u8 unk3{};
+    std::array<u8, 0x41> passphrase{};
+};
+#pragma pack()
+static_assert(sizeof(SfWirelessSettingData) == 0x65, "SfWirelessSettingData has incorrect size.");
+
+#pragma pack(1)
```
Comment (lioncash) [inline src/core/hle/service/nifm/nifm.h:57]: Ditto

File: src/core/hle/service/nifm/nifm.h [hunk 9]
```diff
@@ -14,6 +14,140 @@ class System;
...
+struct IpAddress {
+    u32 address{};
+};
+#pragma pack()
+static_assert(sizeof(IpAddress) == 0x4, "IpAddress has incorrect size.");
+
+#pragma pack(1)
```
Comment (lioncash) [inline src/core/hle/service/nifm/nifm.h:64]: Ditto

File: src/core/hle/service/nifm/nifm.h [hunk 10]
```diff
@@ -14,6 +14,140 @@ class System;
...
+struct SubnetMask {
+    u32 mask{};
+};
+#pragma pack()
+static_assert(sizeof(SubnetMask) == 0x4, "SubnetMask has incorrect size.");
+
+#pragma pack(1)
```
Comment (lioncash) [inline src/core/hle/service/nifm/nifm.h:71]: Ditto

File: src/core/hle/service/nifm/nifm.h [hunk 11]
```diff
@@ -14,6 +14,140 @@ class System;
...
+struct DNSSetting {
+    u8 is_automatic{};
+    u32 primary_dns{};
+    u32 secondary_dns{};
+};
+#pragma pack()
+static_assert(sizeof(DNSSetting) == 0x9, "DNSSetting has incorrect size.");
+
+#pragma pack(1)
```
Comment (lioncash) [inline src/core/hle/service/nifm/nifm.h:97]: Ditto

File: src/core/hle/service/nifm/nifm.h [hunk 12]
```diff
@@ -14,6 +14,140 @@ class System;
...
+    std::array<u8, 0x64> server{};
+    u8 auto_authentication{};
+    std::array<u8, 0x20> user_string{};
+    std::array<u8, 0x20> password{};
+    INSERT_PADDING_BYTES(0x1);
+};
+#pragma pack()
+static_assert(sizeof(ProxySetting) == 0xAA, "ProxySetting has incorrect size.");
+
+#pragma pack(1)
```
Comment (lioncash) [inline src/core/hle/service/nifm/nifm.h:111]: Ditto

File: src/core/hle/service/nifm/nifm.h [hunk 13]
```diff
@@ -14,6 +14,140 @@ class System;
...
+    u8 unk3{};
+    u8 unk4{};
+    INSERT_PADDING_BYTES(0x2);
+    WirelessSettingData wireless_data{};
+    IpSettingData ip_data{};
+};
+#pragma pack()
+static_assert(sizeof(NetworkProfileData) == 0x18E, "NetworkProfileData has incorrect size.");
+
+#pragma pack(1)
```
Comment (lioncash) [inline src/core/hle/service/nifm/nifm.h:136]: Ditto

File: src/core/hle/service/nifm/nifm.cpp [hunk 14]
```diff
@@ -13,6 +13,9 @@
 
 namespace Service::NIFM {
 
+constexpr ResultCode ERR_NO_INTERNET_CONNECTION{ErrorModule::NIFM, 0x12C};
```
Comment (ogniK5377) [inline src/core/hle/service/nifm/nifm.cpp:16]:
```suggestion
constexpr ResultCode ERR_NO_INTERNET_CONNECTION{ErrorModule::NIFM, 300};
```

File: src/core/hle/service/nifm/nifm.cpp [hunk 15]
```diff
@@ -13,6 +13,9 @@
...
+constexpr ResultCode ERR_OBJECT_IS_NULL{ErrorModule::NIFM, 0x15E};
```
Comment (ogniK5377) [inline src/core/hle/service/nifm/nifm.cpp:17]:
```suggestion
constexpr ResultCode ERR_OBJECT_IS_NULL{ErrorModule::NIFM, 350};
```

File: src/core/hle/service/nifm/nifm.cpp [hunk 16]
```diff
@@ -217,7 +220,22 @@ class IGeneralService final : public ServiceFramework<IGeneralService> {
...
         }
     }
+    void GetCurrentIpAddress(Kernel::HLERequestContext& ctx) {
+        LOG_WARNING(Service_NIFM, "(STUBBED) called");
+
+        const auto current_ip_address =
+            network_profile_data.ip_data.address_settings.ip_address.address;
+
+        IPC::ResponseBuilder rb{ctx, 3};
+        if (current_ip_address == 0) {
```
Comment (FearlessTobi) [inline src/core/hle/service/nifm/nifm.cpp:230]: If the ip is always 0, why are we even checking for it? Also where are the error codes taken from?

Comment (DarkLordZach) [reply src/core/hle/service/nifm/nifm.cpp:230]: This check should remain for the future when we use actual IP addresses.

Comment (VolcaEM) [reply src/core/hle/service/nifm/nifm.cpp:230]: The error codes were taken by Ryujinx (https://github.com/Ryujinx/Ryujinx/blob/master/Ryujinx.HLE/HOS/Services/Nifm/ResultCode.cs)

**Comment by VolcaEM** [comment]
Instead of having a faulty stub/implementation, I'd like to have no stub/implementation at all since it could cause issues in various games if stubbed/implemented incorrectly

File: src/core/hle/service/nifm/nifm.cpp (hunk 1 above)
Comment (ogniK5377) [inline src/core/hle/service/nifm/nifm.cpp]:
```suggestion
        const auto current_ip_address = network_profile_data.ip_data.address_settings.ip_address.address;
```

**Comment by VolcaEM** [comment]
Closing because I feel like this alone is not a good change and I am not satisfied with it myself, in the future maybe I'll try implementing some services alongside the necessary structs
About the wrong assert, I'm going to open a separate one-line PR to correct it in the next few hours
If somebody else wants to implement NIFM services in the meantime, obviously feel free to use part of, or the entirety, of these changes

**Pull request closed without merging**